godot-codegen/src/generator/enums.rs

```diff
@@ -5,94 +5,105 @@
+        #bitwise_impls
 
-        if let EnumeratorValue::Enum(ord) = enumerator.value {
-            unique_ords.push(ord);
+        impl crate::builtin::meta::GodotConvert for #name {
+            type Via = #ord_type;
         }
-    }
 
-    let mut derives = vec!["Copy", "Clone", "Eq", "PartialEq", "Hash", "Debug"];
+        impl crate::builtin::meta::ToGodot for #name {
+            fn to_godot(&self) -> Self::Via {
+                <Self as #engine_trait>::ord(*self)
+            }
+        }
 
-    if enum_.is_bitfield {
-        derives.push("Default");
+        impl crate::builtin::meta::FromGodot for #name {
+            fn try_from_godot(via: Self::Via) -> std::result::Result<Self, crate::builtin::meta::ConvertError> {
+                <Self as #engine_trait>::try_from_ord(via)
+                    .ok_or_else(|| crate::builtin::meta::FromGodotError::InvalidEnum.into_error(via))
+            }
+        }
     }
+}
 
-    let derives = derives.into_iter().map(util::ident);
-
-    let index_enum_impl = if enum_.is_bitfield {
-        // Bitfields don't implement IndexEnum.
-        TokenStream::new()
-    } else {
-        // Enums implement IndexEnum only if they are "index-like" (see docs).
-        if let Some(enum_max) = try_count_index_enum(enum_) {
-            quote! {
-                impl crate::obj::IndexEnum for #rust_enum_name {
-                    const ENUMERATOR_COUNT: usize = #enum_max;
-                }
-            }
-        } else {
-            TokenStream::new()
+/// Creates an implementation of `IndexEnum` for the given enum.
+///
+/// Returns `None` if `enum_` isn't an indexable enum.
+fn make_enum_index_impl(enum_: &Enum) -> Option<TokenStream> {
+    let enum_max = enum_.count_index_enum()?;
+    let name = &enum_.name;
+
+    Some(quote! {
+        impl crate::obj::IndexEnum for #name {
+            const ENUMERATOR_COUNT: usize = #enum_max;
         }
-    };
+    })
+}
 
-    let bitfield_ops;
-    let self_as_trait;
-    let engine_impl;
-    let enum_ord_type;
+/// Creates an implementation of the engine trait for the given enum.
+///
+/// This will implement the trait returned by [`Enum::engine_trait`].
+fn make_enum_engine_trait_impl(enum_: &Enum) -> TokenStream {
+    let name = &enum_.name;
+    let engine_trait = enum_.engine_trait();
 
     if enum_.is_bitfield {
-        bitfield_ops = quote! {
-            // impl #enum_name {
-            //     pub const UNSET: Self = Self { ord: 0 };
-            // }
-            impl std::ops::BitOr for #rust_enum_name {
-                type Output = Self;
-
-                fn bitor(self, rhs: Self) -> Self::Output {
-                    Self { ord: self.ord | rhs.ord }
-                }
-            }
-        };
-        enum_ord_type = quote! { u64 };
-        self_as_trait = quote! { <Self as crate::obj::EngineBitfield> };
-        engine_impl = quote! {
-            impl crate::obj::EngineBitfield for #rust_enum_name {
+        quote! {
```

Comment:
Could you re-add the comment here? It was deliberately not removed because I might add it in the future.

Reply:
sure, i'll also add a comment mentioning that

Reply:
Sounds good! :slightly_smiling_face: